platform/src/main/java/org/hillview/storage/CassandraSSTableLoader.java

```diff
@@ -171,137 +230,269 @@ public Schema getSchema(CFMetaData metadata) {
-                Spliterator.IMMUTABLE);
-            Stream<UnfilteredRowIterator> partitions = StreamSupport.stream(splititer, false);
+    private void computeRowCount() {
+        this.rowCount = 0;
+        PartitionColumns.Builder builder = PartitionColumns.builder();
+        ColumnFilter cf = ColumnFilter.selection(builder.build());
 
-            return CassandraSSTableLoader.this.loadColumns(columnToLoad, partitions, this.columnCountToLoad);
+        for (CassandraTokenRange tr : this.tokenRanges) {
+            // Load the partition if the local endpoint is listed as the first endpoint
+            if (tr.endpoints.get(0).equals(this.localEndpoint)) {
+                DataRange range = DataRange.forTokenRange(tr.tokenRange);
+                SSTableReadsListener listener = CassandraSSTableLoader.newReadCountUpdater();
+                ISSTableScanner currentScanner = this.ssTableReader.getScanner(cf, range, false, listener);
+                Spliterator<UnfilteredRowIterator> splititer = Spliterators.spliteratorUnknownSize(currentScanner,
+                        Spliterator.IMMUTABLE);
+                Stream<UnfilteredRowIterator> partitions = StreamSupport.stream(splititer, false)
+                        .filter((p) -> p.hasNext()).filter(r -> r.next() instanceof Row);
+                this.rowCount += partitions.count();
+            }
         }
     }
 
-    public long getNumRows() {
-        ISSTableScanner currentScanner = this.ssTableReader.getScanner();
-        Spliterator<UnfilteredRowIterator> splititer = Spliterators.spliteratorUnknownSize(currentScanner,
-            Spliterator.IMMUTABLE);
-        Stream<UnfilteredRowIterator> partitions = StreamSupport.stream(splititer, false);
-        return partitions.count();
+    public int getRowCount() {
+        if (this.rowCount > Integer.MAX_VALUE || this.rowCount < Integer.MIN_VALUE)
+            throw new RuntimeException("The number of rows exceeds Integer.MAX_VALUE");
+        return (int) this.rowCount;
     }
 
-    /** Instead of checking the columnn' name to find which one to load,
-     * this method uses boolean marker stored at columnToLoad to recognize the needed columns */
-    private List<IAppendableColumn> createColumns(boolean[] columnToLoad, int columnCountToLoad) {
+    /**
+     * Instead of checking the columnn' name to find which one to load, this method
+     * uses boolean marker stored at columnToLoad to recognize the needed columns
+     */
+    private List<IAppendableColumn> createColumns(List<String> names) {
         List<ColumnDescription> cols = Converters.checkNull(this.actualSchema).getColumnDescriptions();
-        List<IAppendableColumn> result = new ArrayList<IAppendableColumn>(columnCountToLoad);
-        int i = 0;
+        List<IAppendableColumn> result = new ArrayList<IAppendableColumn>(names.size());
         for (ColumnDescription cd : cols) {
-            if (columnToLoad[i])
+            if (names.contains(cd.name))
```

Comment:
This could a quadratic cost, which could be large for more than 1000 columns.
Create a set from the List of names.